### ISSUE TITLE
Error: transaction outputs more than it inputs

### ISSUE THREAD

**Issue by gavinharris-dev** [Hyperion-BT/Helios#25]
I have a built a transaction where I am spending Inputs from a Wallet and a Script. The Script UTxO's are paid back to the script and so the value change is ultimately zero.
The Wallet UTxO's are purely for paying Transaction Fees.
The Wallet UTxO Selection process is picking a UTxO that has additional Assets (which should be fine as I'm paying these back as change); however I'm getting the above error that the Transaction is outputting more that it inputs.

I have stepped through the code and think that the issue lies in the `Value::ge` method (reading the comments of the Method it seems that the method is correct, just not the right implementation for here?).

In effect: 
* The Total Lovelace output is less than the total Lovelace input. This is fine.
* The total Assets output is wholly contained by the Transaction Input assets; therefore I think that the 'changeAddress' should be paid out the remaining Assets.

**Total Inputs:**

| Policy | Name | Quantity |
|-------|-------|----------|
| Lovelace | | 9939708590 |
| Lovelace | | 2763219 |
| 1cd7eb4b8635854f55bfaa2651d272264bb82dccdfe67dfb59345520 | 47485431 | 99985 |
| b40c323876955eb432f4360d47d10e647fa6b41081204b0a691cfb89 | 544e4331 | 1 |
| b40c323876955eb432f4360d47d10e647fa6b41081204b0a691cfb89 | 544e4330 | 1 |

**Total Outputs (prior to change calculation)**

| Policy | Name | Quantity |
|-------|-------|----------|
| Lovelace | | 2763219 |
| b40c323876955eb432f4360d47d10e647fa6b41081204b0a691cfb89 | 544e4330 | 1 |

When finalizing the transaction, I provide a Change Address, and so would expect this address to have a Transaction created paying out:

| Policy | Name | Quantity |
|-------|-------|----------|
| Lovelace | | 9939708590~ |
| 1cd7eb4b8635854f55bfaa2651d272264bb82dccdfe67dfb59345520 | 47485431 | 99985 |
| b40c323876955eb432f4360d47d10e647fa6b41081204b0a691cfb89 | 544e4331 | 1 |

~ - Minus transaction fees

Transaction Dump:
```
{
    "body": {
        "inputs": [
            {
                "txId": "a7090340bd529929043c0d96814bd3f8200e798e194f91f500e500dcad948394",
                "utxoIdx": "1",
                "origOutput": {
                    "address": {
                        "hex": "00d006eb7783e8c93160b2bab287bc8a6f069e9e690cd82bc0b52a8c31730d805b6a2cf67998f0fcd070ce0b6e85957fd758cf0ae348d265eb",
                        "bech32": "addr_test1qrgqd6mhs05vjvtqk2at9pau3fhsd857dyxds27qk54gcvtnpkq9k63v7eue3u8u6pcvuzmwsk2hl46ceu9wxjxjvh4sj4drgd"
                    },
                    "value": {
                        "lovelace": "9939708590",
                        "assets": {
                            "1cd7eb4b8635854f55bfaa2651d272264bb82dccdfe67dfb59345520": {
                                "47485431": "99985"
                            },
                            "b40c323876955eb432f4360d47d10e647fa6b41081204b0a691cfb89": {
                                "544e4331": "1"
                            }
                        }
                    },
                    "datum": null,
                    "refScript": null
                }
            },
            {
                "txId": "43cd25ddc7bdb9bf823bb3552713935a79353585ddb67ab89458c2257800ae8b",
                "utxoIdx": "0",
                "origOutput": {
                    "address": {
                        "hex": "10b316d8aa101fc0546af6e00ed0b75050e550dda4442346a8a1239b21dabfdce2c315fedb09e4ba39dbc4345866257eee43d9302e3e4b4013",
                        "bech32": "addr_test1zze3dk92zq0uq4r27msqa59h2pgw25xa53zzx34g5y3ekgw6hlww9sc4lmdsne9688dugdzcvcjhamjrmyczu0jtgqfs2fvtx2"
                    },
                    "value": {
                        "lovelace": "2763219",
                        "assets": {
                            "b40c323876955eb432f4360d47d10e647fa6b41081204b0a691cfb89": {
                                "544e4330": "1"
                            }
                        }
                    },
                    "datum": {
                        "hash": "ded341b0bbf8c1ecebbaa7a8a5c91d663177aff4891c20d0cb898c1f377e5593",
                        "cbor": "d8799fa1581cd006eb7783e8c93160b2bab287bc8a6f069e9e690cd82bc0b52a8c31194e20a0581cd006eb7783e8c93160b2bab287bc8a6f069e9e690cd82bc0b52a8c311a0876bf80d8799f581cb40c323876955eb432f4360d47d10e647fa6b41081204b0a691cfb8944544e4330ff1b000001842c323a0018321a004c4b40ff",
                        "schema": {
                            "constructor": 0,
                            "fields": [
                                {
                                    "map": [
                                        {
                                            "k": {
                                                "bytes": "d006eb7783e8c93160b2bab287bc8a6f069e9e690cd82bc0b52a8c31"
                                            },
                                            "v": {
                                                "int": 20000
                                            }
                                        }
                                    ]
                                },
                                {
                                    "map": []
                                },
                                {
                                    "bytes": "d006eb7783e8c93160b2bab287bc8a6f069e9e690cd82bc0b52a8c31"
                                },
                                {
                                    "int": 142000000
                                },
                                {
                                    "constructor": 0,
                                    "fields": [
                                        {
                                            "bytes": "b40c323876955eb432f4360d47d10e647fa6b41081204b0a691cfb89"
                                        },
                                        {
                                            "bytes": "544e4330"
                                        }
                                    ]
                                },
                                {
                                    "int": 1667188800000
                                },
                                {
                                    "int": 50
                                },
                                {
                                    "int": 5000000
                                }
                            ]
                        }
                    },
                    "refScript": null
                }
            }
        ],
        "outputs": [
            {
                "address": {
                    "hex": "10b316d8aa101fc0546af6e00ed0b75050e550dda4442346a8a1239b21730d805b6a2cf67998f0fcd070ce0b6e85957fd758cf0ae348d265eb",
                    "bech32": "addr_test1zze3dk92zq0uq4r27msqa59h2pgw25xa53zzx34g5y3ekgtnpkq9k63v7eue3u8u6pcvuzmwsk2hl46ceu9wxjxjvh4sp2tnz4"
                },
                "value": {
                    "lovelace": "2763219",
                    "assets": {
                        "b40c323876955eb432f4360d47d10e647fa6b41081204b0a691cfb89": {
                            "544e4330": "1"
                        }
                    }
                },
                "datum": {
                    "inlineCbor": "d8799fa1581cd006eb7783e8c93160b2bab287bc8a6f069e9e690cd82bc0b52a8c31194e20a0581cd006eb7783e8c93160b2bab287bc8a6f069e9e690cd82bc0b52a8c311a08a48640d8799f581cb40c323876955eb432f4360d47d10e647fa6b41081204b0a691cfb8944544e4330ff1b00000184403bc8df18321a004c4b40ff",
                    "inlineSchema": {
                        "constructor": 0,
                        "fields": [
                            {
                                "map": [
                                    {
                                        "k": {
                                            "bytes": "d006eb7783e8c93160b2bab287bc8a6f069e9e690cd82bc0b52a8c31"
                                        },
                                        "v": {
                                            "int": 20000
                                        }
                                    }
                                ]
                            },
                            {
                                "map": []
                            },
                            {
                                "bytes": "d006eb7783e8c93160b2bab287bc8a6f069e9e690cd82bc0b52a8c31"
                            },
                            {
                                "int": 145000000
                            },
                            {
                                "constructor": 0,
                                "fields": [
                                    {
                                        "bytes": "b40c323876955eb432f4360d47d10e647fa6b41081204b0a691cfb89"
                                    },
                                    {
                                        "bytes": "544e4330"
                                    }
                                ]
                            },
                            {
                                "int": 1667524970719
                            },
                            {
                                "int": 50
                            },
                            {
                                "int": 5000000
                            }
                        ]
                    }
                },
                "refScript": null
            }
        ],
        "fee": "0",
        "lastValidSlot": null,
        "firstValidSlot": null,
        "minted": null,
        "scriptDataHash": null,
        "collateral": [
            {
                "txId": "10e44e9cb480b27347bef38c5b5e3a6ee0ccd20e43c4ec99ba1e6fa05483a3f2",
                "utxoIdx": "0",
                "origOutput": {
                    "address": {
                        "hex": "00d006eb7783e8c93160b2bab287bc8a6f069e9e690cd82bc0b52a8c31730d805b6a2cf67998f0fcd070ce0b6e85957fd758cf0ae348d265eb",
                        "bech32": "addr_test1qrgqd6mhs05vjvtqk2at9pau3fhsd857dyxds27qk54gcvtnpkq9k63v7eue3u8u6pcvuzmwsk2hl46ceu9wxjxjvh4sj4drgd"
                    },
                    "value": {
                        "lovelace": "5000000",
                        "assets": {}
                    },
                    "datum": null,
                    "refScript": null
                }
            }
        ],
        "signers": [
            "d006eb7783e8c93160b2bab287bc8a6f069e9e690cd82bc0b52a8c31"
        ],
        "metadataHash": null,
        "refInputs": []
    },
    "witnesses": {
        "signatures": [],
        "datums": [
            "0{{#d006eb7783e8c93160b2bab287bc8a6f069e9e690cd82bc0b52a8c31: 20000}, {}, #d006eb7783e8c93160b2bab287bc8a6f069e9e690cd82bc0b52a8c31, 142000000, 0{#b40c323876955eb432f4360d47d10e647fa6b41081204b0a691cfb89, #544e4330}, 1667188800000, 50, 5000000}"
        ],
        "redeemers": [
            {
                "data": "1{1667524970719, 145000000}",
                "exUnits": {
                    "mem": "0",
                    "cpu": "0"
                },
                "type": 0,
                "typeName": "spending",
                "inputIndex": -1
            }
        ],
        "scripts": [
            "sdfds"
        ],
        "refScripts": []
    },
    "valid": false,
    "metadata": null
}
```

**Comment by christianschmitz** [comment]
Something else is happening here. "valid" should be "true" if the transaction is properly finalized.
Also the "witnesses.scripts" doesn't contain a valid script, and "fee" is still "0", "inputIndex" in "redeemers" shouldn't be -1.

Did you call `await tx.finalize()`, or another method?

**Comment by gavinharris-dev** [comment]
Yeah the error is being thrown from the `await tx.finalize()` call.

I have it working now by running the following 'balancing' process:

```
    const balanceAssets = addedWalletUTxOs
      .map((wUtxo) => wUtxo.value.assets)
      .reduce((allAssets, assets) => allAssets.add(assets), new Assets());

    if (!balanceAssets.isZero()) {
      // Create a TXout to the Change Address with these Unbalanced Assets
      const balanceOutput = new TxOutput(
        walletAddress,
        new Value(BigInt(0), balanceAssets)
      );
      balanceOutput.correctLovelace(networkParams);

      tx.addOutput(balanceOutput);
    }
```

Where `addedWalletUTxOs` is an array of UTxO's used as Inputs to the Transaction. So this is balancing the 'MultiAsset' inputs that are added from the Wallet.

**Comment by gavinharris-dev** [comment]
Also the `tx.dump()` output is from immediately prior to the call to Finalize; hence the fee being zero (and the transaction not balancing).

**Comment by gavinharris-dev** [comment]
I'll see if I can put together a gist.

**Comment by christianschmitz** [comment]
> Yeah the error is being thrown from the await tx.finalize() call.

What is the error?

**Comment by gavinharris-dev** [comment]
I am failing to replicate this issue outside of the project. Seems rather odd but its either something that has been fixed (I am still running 0.8.8) or (and this is FAR more likely) there is an idiot in Queensland that has done something a little odd and needed to write a work around. Perhaps it will fix up when I refactor.

**Comment by christianschmitz** [comment]
Helios' error messages could probably be improved drastically
